### PR TITLE
Added support for NVMe

### PR DESCRIPTION
**Pull request by alexth4ef9**
Support for NVMe SSD. Should fix the SSD part of issue #963 (the mentioned Samsung SM951 is a NVMe SSD)

Tested on Win10 64-bit, SSD is an Intel SSD 750.
ohm_nvme
Part of the Report
> NVMeGeneric
> 
> Drive name: INTEL SSDPEDMW012T4
> Firmware version: 8EV10174
> 
> PCI Vendor ID: 0x8086
> IEEE OUI Identifier: 0x5cd2e4
> Total NVM Capacity: 0
> Unallocated NVM Capacity: 0
> Controller ID: 0
> Number of Namespaces: 1
> Namespace 1 Size: 1200243695616
> Namespace 1 Capacity: 1200243695616
> Namespace 1 Utilization: 1200243695616
> Namespace 1 LBA Data Size: 512
> Critical Warning: -
> Temperature: 34 Celsius
> Available Spare: 100%
> Available Spare Threshold: 10%
> Data Units Read: 21976904
> Data Units Written: 13212023
> Host Read Commands: 301480076
> Data Write Commands: 215624853
> Controller Busy Time: 0
> Power Cycles: 599
> Power On Hours: 4707
> Unsafe Shutdowns: 0
> Media Errors: 0
> Number of Error Information Log Entries: 0
> Warning Composite Temperature Time: 0
> Critical Composite Temperature Time: 0

I had to split AbstractHarddrive.cs into AbstractHarddrive and AbstractStorage classes, because NVMe doesn't fit the SMART model. AbstractStorage.CreateInstanstance will create AbstractHarddrive for SATA/ATA drives or NVMeGeneric for NVMe drives.

### CODE REVIEW

**Comment by DarthDestroyer** [comment]
Good Addition! Well done!

Only a small Thing:

> Host Read Commands: 301480076
> Data Write Commands: 215624853

You switched between Host / Data Namings, is that correct?
With IntelSSD you get "Host Writes / Host Reads" - maybe a better naming Style to use the same?

**Comment by alexth4ef9** [comment]
Thanks for pointing this out. Should be Host Write Commands (even the variable is named this). In the report is another mistake Percentage Used is missing (but shown as sensor).  Will fix both things later.

**Comment by KrzysFR** [comment]
I tried this patch on a Samsung 960 EVO and 960 Pro, and both drives are NOT detected at all.

I tracked the issue in `NVMeGeneric.GetDeviceInfo(...)` where the `serial` argument seems wonky:
- The `serial` argument does not seem right at all, with values that look like `"0012_3456_789A_BCDE.`" (with `_` and a trailing `.`)
- The `serialToInfo` dictionary has the correct serials for my drives (that look like `"S3ETNxxxx"`).

This means that the serial is not found in the table, and the method returns null.

I looked at the binary blob that is parsed in `StorageInfoImpl.ctor()` and it contains the incorrect "serial", so this looks like that the data return by `DeviceIoControl(...)` in `WindowsStorage.GetStorageInfo(..)` does not contain what we expect.

What can I do to help test this?

_On Win10, MSI Z270, i7-7700K_

**Comment by alexth4ef9** [comment]
Using a simple search for getting the next NVMe Controller (Logical a NVMe disk is a SCSI controller with one attached disk).
I tried to use the new Windows 10 STORAGE_QUERY_PROPERTY for NVMe (https://msdn.microsoft.com/en-us/library/windows/desktop/mt718131(v=vs.85).aspx), but it doesn't work.

**Comment by KrzysFR** [comment]
With your last commit, I'm getting this error on start:
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at OpenHardwareMonitor.Hardware.HDD.NVMeGeneric.<>c__DisplayClass10_0.<CreateSensors>b__6(NVMeHealthInfo health)
   at OpenHardwareMonitor.Hardware.HDD.NVMeGeneric.NVMeSensor.Update(NVMeHealthInfo health)
   at OpenHardwareMonitor.Hardware.HDD.NVMeGeneric.UpdateSensors()
   at OpenHardwareMonitor.Hardware.HDD.AbstractStorage.Update()
   at OpenHardwareMonitor.GUI.UpdateVisitor.VisitHardware(IHardware hardware)
   at OpenHardwareMonitor.Hardware.Computer.Traverse(IVisitor visitor)
   at OpenHardwareMonitor.GUI.MainForm.timer_Tick(Object sender, EventArgs e)
   at System.Windows.Forms.Timer.OnTick(EventArgs e)
   at System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```
In `NVMeGeneric.CreateSensors()`, the update handler is called with `i == 8`, but  `TemperatureSensors.Length == 8`.

```
for (int i=0; i<log.TemperatureSensors.Length; i++) {
    if (log.TemperatureSensors[i] > short.MinValue) {
        AddSensor("Temperature", i + 1, true, SensorType.Temperature,
            (health) => health.TemperatureSensors[i]);  // <-- gets called with i == 8 !
    }
}
```

I think this is the classical issue where you are hoisting the value of `i` in the closure, which gets updated by the `i++` in the loop, so when the callback fires, its value is 8. Copying the value of `i` into a temporary value `idx` and using that fixes the issue:
```
for (int i=0; i<log.TemperatureSensors.Length; i++) {
    if (log.TemperatureSensors[i] > short.MinValue) {
        int idx = i;
        AddSensor("Temperature", i + 1, true, SensorType.Temperature,
            (health) => health.TemperatureSensors[idx]);
    }
}
```
_It was my impression that this particular issue was fixed in recent versions of C# (the compiler would create the tmp variable for you), but for some reason it's not doing it in this case? weird..._

Now, I can see both my disks, and the values seems sensible (temperature 37-42°C)

**Comment by alexth4ef9** [comment]
Thanks for testing it. It's probably my dev environment, that compiler bug didn't showed up. I'm using Sharpdevelop and .NET 4.6.2 SDK. Even compiling for .NET 2.0, the compiler is C# 5.0.

**Comment by sebastienwarin** [comment]
Hello,

I tried this patch on a Samsung SSD 960 EVO 250GB (on a ASRock Z390M w/ Core i5 8400 - 8th generation).

The WindowsStorage.GetStorageInfo() detect my drive #0 with the BusTypeNvme then NVMeGeneric.CreateInstance is called with the correct StorageInfo (Serial: 0025_3858_71B0_DBC8. Name/Product : Samsung SSD 960 EVO 250GB).

But in the WindowsNVMeSmart.GetInfo() I have an exception here  : 
`NVMeIdentifyControllerData data = ReadPassThrough<NVMeIdentifyControllerData>(NVMePassThroughOpcode.AdminIdentify, 0x00000000, 0x000000001, out rawData);`

`System.ComponentModel.Win32Exception (0x80004005): The request could not be performed because of an I/O device error
   at OpenHardwareMonitor.Hardware.HDD.WindowsNVMeSmart.ReadPassThrough[T](NVMePassThroughOpcode opcode, UInt32 nsid, UInt32 cdw10, Byte[]& rawData)
   at OpenHardwareMonitor.Hardware.HDD.WindowsNVMeSmart.GetInfo()`

Can you help me ?

**Comment by donserdal** [comment]
Works with a "Samsung SSD 970 EVO 500GB" as well!

**Comment by louitrilobite** [comment]
When I use this pull request my Samsung SSD 970 EVO 500GB shows up nicely (it doesn't without) but my Samsung SSD 860 EVO 500GB looses its temperature readings as do my normal HD drives.

Without this pull request...



With...



Is there something I can do to have my other drives show their info?

Thanks.

**Comment by aster94** [comment]
> Works with a "Samsung SSD 970 EVO 500GB" as well!

perfectly working also for me!

**Comment by Sosenek** [comment]
Samsung m2 SSD 970 Pro 512GB also working. OHM 8.0.5 Alpha
ohm

**Comment by DevilsJoker** [comment]
With these changes, which works fine with my other PC's, I can't find any hard disks on my Surface Book 2, not even a generic hard disk?

image

The disk is a SAMSUNG MZFLW256HEHP-000MV NVMe.

**Comment by DevilsJoker** [comment]
I can't make this work with the standard windows 10 nvme driver. Am I the only one, or is this a general thing?

I have a Corsair and a WD Black disk, and these don't have their own drivers..

**Comment by jpstotz** [comment]
On my system the NVMe drive is also not shown (Samsung 970 Evo Plus NVMe, Windows stornvme.sys driver). Only the SATA SSD and my HDD is shown.

It fails to retrieve any info on the NVMe drive at this point:
https://github.com/alexth4ef9/openhardwaremonitor/blob/nvme-patch/Hardware/HDD/WindowsNVMeSmart.cs#L481

Error is: `Error Code 1117 - The request could not be performed because of an I/O device error`

**Comment by jpstotz** [comment]
On my system the NVMe drive is also not shown (Samsung 970 Evo Plus NVMe). Only the SATA SSD and my HDD is shown.

It fails to retrieve any info on the NVMe drive at this point:
https://github.com/alexth4ef9/openhardwaremonitor/blob/nvme-patch/Hardware/HDD/WindowsNVMeSmart.cs#L481

Error is: `Error Code 1117 - The request could not be performed because of an I/O device error`

Only after installing the Samsung NVMe driver the code of this PR works.

**Comment by lprocx** [comment]
Since this PR is from 2017, The features of the last 3.5 years are missing of course. Tried the branch and information of my NVMe drives are shown perfectly. I took the changes from this branch and tried implementing it into the master again. I only have Samsung NVMe drives, but everything is working as it should be. See the image below for my results.

I'm being honest here by thinking that I'm a fine programmer in my expertises, but a lot of the code was still new to me. If a new PR is needed I can make one, but I'm unable to judge if all of my changes are 100% correct and ready to be merged to the master. 

The project can be found at [https://github.com/lprocx/openhardwaremonitor](https://github.com/lprocx/openhardwaremonitor)

If we do want this into master I will make a pullrequest. With some feedback it should be good to go.

Of course all credit of this goes to @alexth4ef9

ohw-nvme

**Comment by jpstotz** [comment]
@lprocx Thanks for making it possible to use the NVMe patch with a recent version of OpenHardwareMonitor. I tried to to the same but failed. Hence I understand that this was not a simple job of merging some code, as you wrote this was more a reimplementation. 

I tested your version and in general in works. The only problem I recognized is that the NVMe drive disappears when OHM is running and you hibernate the system (e.g. suspend to disc). Afterwards the NVMe entry was gone, but I have to test that in detail.

**Comment by lprocx** [comment]
> @lprocx Thanks for making it possible to use the NVMe patch with a recent version of OpenHardwareMonitor. I tried to to the same but failed. Hence I understand that this was not a simple job of merging some code, as you wrote this was more a reimplementation.
> 
> I tested your version and in general in works. The only problem I recognized is that the NVMe drive disappears when OHM is running and you hibernate the system (e.g. suspend to disc). Afterwards the NVMe entry was gone, but I have to test that in detail.

Glad it helped. I have no knowledge how OHM handles hibernation, but I'm sure somebody else can tackle this issue. Any progress on your side so far?

**Comment by jpstotz** [comment]
> Glad it helped. I have no knowledge how OHM handles hibernation, but I'm sure somebody else can tackle this issue. Any progress on your side so far?

I am not sure. OpenHardwareMonitor seems to have some problems in general with energy saving states. When I use the release version 0.9.5, start it, enter Suspend to RAM, and finally resume the temperature entries on my two SATA discs are gone. 
Only if I perform a "Reset" via menu the temperature items reappear.
The same problem happens when I use your branch.

I even ran your version from within Visual Studio while entering energy saving state, but I could not observe any error, exception or anything unusual. I would assume that there are just code parts that are not executed again after resume. 

The "fun" part is that if I use the old branch this PR bases on I don't have any problems after resuming from energy saving (except occasionally OHM crashes after resuming from energy saving).

So in conclusion I would say that your branch works as good as the existing master branch. From my perspective there are no obstacles that would prevent merging your code (unfortunately my direct influence on this project is next to zero on this project).

**Comment by jpstotz** [comment]
@lprocx In my opinion you should make a PR for your code. It is better to have something that works 99% instead of something that does not work at all when it comes to NVMe drives. 

Afterwards we still have to convince the maintainer of this project to merge it. Looking at the list of open PR requests the maintainer(s) does not seem to care about this project much :(

**Comment by lprocx** [comment]
> @lprocx In my opinion you should make a PR for your code. It is better to have something that works 99% instead of something that does not work at all when it comes to NVMe drives.
> 
> Afterwards we still have to convince the maintainer of this project to merge it. Looking at the list of open PR requests the maintainer(s) does not seem to care about this project much :(

@jpstotz I will try to fix a PR this week (little busy). I'm not familiar enough with the code from this project, but if more maintainers are needed I would try and learn the code a bit more and help where I can.

**Comment by lprocx** [comment]
@jpstotz PR Opened: https://github.com/openhardwaremonitor/openhardwaremonitor/pull/1402
